### PR TITLE
Reduce OP

### LINKED ISSUE #1073: Update Reduce operator

### PR DESCRIPTION
**Pull request by lambda7xx**
**Description of changes:**
- update the Reduce OP

**Related Issues:**

Linked Issues:
- Issue #1073 

Issues closed by this PR:
- Closes #1073

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1118)

### CODE REVIEW

File: lib/runtime/src/ops/reduce.cc [hunk 1]
```diff
@@ -45,391 +48,567 @@ ReduceParams Reduce::get_params() const {
...
+static DeviceSpecific<ReducePerDeviceState>
+    init_task_impl(TaskArgumentAccessor const &acc) {
+  PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
+  auto attrs = acc.get_argument<ReduceAttrs>(ATTRS);
+  OperatorType = attrs.op_type;
+  // Note: How to set the reduction size?
+  ffTensorDescriptor_t inputTensor;
+  ffTensorDescriptor_t outputTensor;
+  ffReduceTensorDescriptor_t reduceDesc;
+  size_t reduction_size DeviceSpecific<ReducePerDeviceState> per_device_state =
```
Comment (lambda7xx) [inline lib/runtime/src/ops/reduce.cc]: how to get `reduction_size`?

File: lib/runtime/src/ops/reduce.cc [hunk 2]
```diff
@@ -45,391 +48,182 @@ ReduceParams Reduce::get_params() const {
...
+static DeviceSpecific<ReducePerDeviceState>
+    init_task_impl(TaskArgumentAccessor const &acc) {
+  PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
+  auto attrs = acc.get_argument<ReduceAttrs>(ATTRS);
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+
+  OperatorType = attrs.op_type;
+  // Note: How to set the reduction size?
+  size_t reduction_size ;
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/reduce.cc:82]: I think `reduction_size = input.shape.get_volume() / output.shape.get_volume();`

File: lib/runtime/src/ops/reduce.cc [hunk 3]
```diff
@@ -45,391 +48,182 @@ ReduceParams Reduce::get_params() const {
...
-    Layer const *layer,
-    std::vector<ParallelTensor> const &inputs) {
-  std::vector<int> axes;
-  long long value;
-  layer->get_int_vector_property("legion_axes", axes);
-  layer->get_int_property("keepdims", value);
-  bool keepdims = value;
-  return new Reduce(
-      model, layer->op_type, inputs[0], axes, keepdims, layer->name);
+plate<> void register_task<TRANSPOSE_INIT_TASK_ID>() {
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/reduce.cc:98]:
```suggestion
template<> void register_task<REDUCE_INIT_TASK_ID>() {
```

**Review by reyna-abhyankar — CHANGES_REQUESTED** [review]

File: lib/runtime/src/ops/reduce.cc (hunk 2 above)
Comment (lambda7xx) [reply lib/runtime/src/ops/reduce.cc:82]: @lockshaw  what do you think about this?

File: lib/runtime/src/ops/reduce.cc (hunk 3 above)
Comment (lambda7xx) [reply lib/runtime/src/ops/reduce.cc:98]: thanks

File: lib/runtime/src/ops/reduce.cc (hunk 2 above)
Comment (reyna-abhyankar) [reply lib/runtime/src/ops/reduce.cc]: bump @lockshaw

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/reduce.cc` line 82 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1118#gh-1341913904:-NgzMG7z38PsoKnCUHjX:b-xj8s5t) ([raw file](https://github.com/flexflow/FlexFlow/blob/bb542fdcb36e8cc41781c4248b103d97005188af/lib/runtime/src/ops/reduce.cc#L82)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

bump @lockshaw 
</blockquote></details>

@lockshaw

**Review by reyna-abhyankar — APPROVED** [review]

**Pull request merged**